fsmonitor.c

```diff
@@ -32,17 +33,29 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 	uint32_t ewah_size;
 	struct ewah_bitmap *fsmonitor_dirty;
 	int ret;
+	struct strbuf last_update = STRBUF_INIT;
 
 	if (sz < sizeof(uint32_t) + sizeof(uint64_t) + sizeof(uint32_t))
 		return error("corrupt fsmonitor extension (too short)");
 
 	hdr_version = get_be32(index);
 	index += sizeof(uint32_t);
-	if (hdr_version != INDEX_EXTENSION_VERSION)
+	if (hdr_version == INDEX_EXTENSION_VERSION1) {
+		/*
+		 * Need have a blank last_update to get a valid token
+		 * when running the next query against the fsmonitor
+		 */
+		index += sizeof(uint64_t);
+	} else if (hdr_version == INDEX_EXTENSION_VERSION2) {
+		strbuf_addstr(&last_update, index);
+		index += last_update.len;
+	} else {
+		strbuf_release(&last_update);
 		return error("bad fsmonitor version %d", hdr_version);
+	}
 
-	istate->fsmonitor_last_update = get_be64(index);
-	index += sizeof(uint64_t);
+	istate->fsmonitor_last_update = strbuf_detach(&last_update, NULL);
+	strbuf_release(&last_update);
```

Comment:
This `strbuf_release()` isn't needed if you just `detach`ed.

Reply:
`strbuf_detach` does a `strbuf_init` so shouldn't it still need to be released?  Or is it because the `fsmonitor_last_update` has the memory that needs to be freed, release doesn't need to happen?
```
/**
 * Detach the string from the strbuf and returns it; you now own the
 * storage the string occupies and it is your responsibility from then on
 * to release it with `free(3)` when you are done with it.
 *
 * The strbuf that previously held the string is reset to `STRBUF_INIT` so
 * it can be reused after calling this function.
 */
char *strbuf_detach(struct strbuf *sb, size_t *sz)
{
	char *res;
	strbuf_grow(sb, 0);
	res = sb->buf;
	if (sz)
		*sz = sb->len;
	strbuf_init(sb, 0);
	return res;
}
```